File: 3625.java
import java.io.BufferedReader;
 import java.io.IOException;
 import java.io.InputStreamReader;

 public class Main {

     public static void main(String[] args) throws IOException {
         BufferedReader read = new BufferedReader(new InputStreamReader(
                 System.in));
         String[] s = read.readLine().split(" ");
         int n = Integer.parseInt(s[0]);
         int m = Integer.parseInt(s[1]);
         int p1, p2;
         double[][] a = new double[n][n];
         int[][] p = new int[n][2];
         for (int i = 0; i < n; i++) {
             s = read.readLine().split(" ");
             p[i][0] = Integer.parseInt(s[0]);
             p[i][1] = Integer.parseInt(s[1]);
         }
         for (int i = 0; i < n; i++) {
             for (int j = i + 1; j < n; j++) {
                 a[i][j] = a[j][i] = Math.sqrt(Math.pow(p[i][0] - p[j][0], 2)
                         + Math.pow(p[i][1] - p[j][1], 2));
             }
         }
         for (int i = 0; i < m; i++) {
             s = read.readLine().split(" ");
             p1 = Integer.parseInt(s[0]);
             p2 = Integer.parseInt(s[1]);
             a[p1 - 1][p2 - 1] = a[p2 - 1][p1 - 1] = 0;
         }
         System.out.printf("%.2f", prim(a, n));
     }

     public static double prim(double[][] a, int count) {
         double sum = 0;
         int i, j, k;
         double[] lowcost = new double[count];
         double[] closeset = new double[count];
         boolean[] used = new boolean[count];
         for (i = 0; i < count; i++) {
             lowcost[i] = a[0][i];
             closeset[i] = 0;
             used[i] = false;
         }
         used[0] = true;
         for (i = 1; i < count; i++) {
             j = 0;
             while (used[j]) {
                 j++;
             }
             for (k = 0; k < count; k++) {
                 if ((!used[k]) && (lowcost[k] < lowcost[j])) {
                     j = k;
                 }
             }
             sum += lowcost[j];
             used[j] = true;
             for (k = 0; k < count; k++) {
                 if (!used[k] && (a[j][k] < lowcost[k])) {
                     {
                         lowcost[k] = a[j][k];
                         closeset[k] = j;
                     }
                 }
             }
         }
         return sum;
     }
}

File: 3625.cc
#include <iostream>
#include <math.h>
#include <cstdio>
#define MAX 100000000
using namespace std;
double a[1001][1001],w[1001];
bool b[1001];
double x[1001],y[1001];
int n;
void Prim(){
 int i,j,t;
 double tmp,len=0;
 for(i=1;i<=n;i++)
 {
 w[i]=a[1][i];
 b[i]=false;
 }
 b[1]=true;
 for(i=2;i<=n;i++)
 {
 t=1;
 tmp=MAX;
 for(j=2;j<=n;j++)
 {
 if(!b[j]&&w[j]<tmp)
 {
 t=j;
 tmp=w[j];
 }
 }
 len+=tmp;
 b[t]=true;
 for(j=2;j<=n;j++)
 {
 if(!b[j]&&w[j]>a[t][j])
 w[j]=a[t][j];
 }
 }
 printf("%.2f\n",len);

}
int main()
{
 int m,i,j;
 scanf("%d%d",&n,&m);
 for(i=1;i<=n;i++)
 cin>>x[i]>>y[i];
 for(i=1;i<=n;i++)
 for(j=1;j<=i;j++)
 {
 a[i][j]=sqrt((x[i]-x[j])*(x[i]-x[j])+(y[i]-y[j])*(y[i]-y[j]));
 a[j][i]=a[i][j];
 }
 while(m--)
 {
 scanf("%d%d",&i,&j);
 a[i][j]=0;
 a[j][i]=0;
 }
 Prim();
 return 0;
}


